### ISSUE TITLE
torch.fft.fft/ifft give incorrect output on contiguous permuted tensors

### ISSUE THREAD

**Issue by ad12** [pytorch/pytorch#71321]
### 🐛 Describe the bug

`torch.fft.fft2` and `torch.fft.ifft2` do not produce the correct output for permuted tensors when the fft/ifft are taken along the same corresponding dimensions.

Permuting tensors should not affect the fft/ifft operations as long as the operations are performed along same dimensions that correspond to the new permuted tensors. However, torch.fft.fft and torch.fft.ifft give incorrect ouptuts on permuted tensors. We also find that the norm= argument and the dtype of the underlying tensor impacts the magnitude of the error.

In the example below, `X1` is a complex 4D tensor with dimensions corresponding to `(B, H, W, C)`. `X2` is the permutation of `X1` such that dimensions are ordered as `(B, C, H, W)`. Taking the ifft2 over `dim=(1,2)` in `X1` should be equivalent to taking the ifft2 over `dim=(-2,-1)` in `X2` - i.e. the error should be non-zero.

Note, if the `.contiguous()` is removed, the error is 0. Thus, the bug may be related to the memory layout for complex tensors.

We discuss some general observations related to the effect of the norm and dtype in the section below. The code below and these observations are also available at [this colab link](https://colab.research.google.com/drive/1Qi8QKjglP-BgsusqSuv-998N_20AwQsZ?authuser=1#scrollTo=hVOpjnNiNeq5).

```python
import torch

torch.manual_seed(0)

# Generate random complex tensor.
X1 = torch.randn((1, 100, 50, 8), dtype=torch.complex64)  # (B, H, W, C)

# 1. Take the centered 2D ifft along dim=(1,2).
x1 = torch.fft.ifft2(X1, norm="ortho", dim=(1, 2))

# 2. Permute X1 and take centered 2D ifft.
# This fails - the error is non-zero
X2 = X1.permute((0, 3, 1, 2)).contiguous()  # (B, C, H, W)
x2 = torch.fft.ifft2(X2, norm="ortho", dim=(-2, -1))
x2 = x2.permute(0, 2, 3, 1)  # (B, H, W, C)

print("x1-x2 error: ", torch.abs(x1 - x2).sum())
```

Output:

```python
x1-x2 error:  tensor(0.0048)
```

## General Observations
These experiments show that the `norm` argument and `dtype` of the tensor impact the magnitude of the error.

We observe that the error is much smaller for 128-bit complex tensors, which may indicate that this bug could be related to floating point errors. However, `torch.complex64` is a widely used dtype and errors of the magnitude seen below are much larger than they should be.

It is a bit unclear why different normalization methods would impact the error, as the normalization constant should be the same regardless of if the tensor if permuted or not.

```python
def runner(norm="ortho", dtype=torch.complex64):
  # This is a copy of the code above
  torch.manual_seed(0)

  X1 = torch.randn((1, 100, 50, 8), dtype=dtype)  # (B, H, W, C)

  # 1. Take the centered 2D ifft along dim=(1,2).
  x1 = torch.fft.ifft2(X1, norm=norm, dim=(1, 2))

  # 2. Permute X1 and take centered 2D ifft.
  # This fails - the error is non-zero
  X2 = X1.permute((0, 3, 1, 2)).contiguous()  # (B, C, H, W)
  x2 = torch.fft.ifft2(X2, norm=norm, dim=(-2, -1))
  x2 = x2.permute(0, 2, 3, 1)  # (B, H, W, C)

  print(f"norm={norm}, dtype={dtype} error: ", torch.abs(x1 - x2).sum())


for dtype in [torch.complex64, torch.complex128]:
  for norm in ["forward", "ortho", "backward"]:
    runner(norm, dtype)
```

### Versions

**Note:** The environment details below are collected from a local machine (not Colab). The colab environment details can be found at [this colab link](https://colab.research.google.com/drive/1Qi8QKjglP-BgsusqSuv-998N_20AwQsZ?authuser=1#scrollTo=hVOpjnNiNeq5).

Collecting environment information...
PyTorch version: 1.10.1
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.3 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.17

Python version: 3.7.11 (default, Jul 27 2021, 14:32:16)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.4.0-91-generic-x86_64-with-debian-bullseye-sid
Is CUDA available: True
CUDA runtime version: 9.0.176
GPU models and configuration: GPU 0: TITAN Xp
Nvidia driver version: 460.91.03
cuDNN version: Probably one of the following:
/usr/local/cuda-10.0/lib64/libcudnn.so.7.3.0
/usr/local/cuda-10.1/targets/x86_64-linux/lib/libcudnn.so.7.6.5
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn.so.8.0.0
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.0.0
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.0.0
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.0.0
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.0.0
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.0.0
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.0.0
/usr/local/cuda-8.0/lib64/libcudnn.so.5.1.10
/usr/local/cuda-9.0/lib64/libcudnn.so.7.1.2
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.21.2
[pip3] pytorch-lightning==1.5.7
[pip3] torch==1.10.1
[pip3] torchmetrics==0.5.1
[pip3] torchvision==0.11.2
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               10.2.89              hfd86e86_1  
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2021.3.0           h06a4308_520  
[conda] mkl-service               2.4.0            py37h7f8727e_0  
[conda] mkl_fft                   1.3.1            py37hd3c417c_0  
[conda] mkl_random                1.2.2            py37h51133e4_0  
[conda] mypy-extensions           0.4.3                    pypi_0    pypi
[conda] numpy                     1.20.3                   pypi_0    pypi
[conda] numpy-base                1.21.2           py37h79a1101_0  
[conda] pytorch                   1.10.1          py3.7_cuda10.2_cudnn7.6.5_0    pytorch
[conda] pytorch-lightning         1.5.7                    pypi_0    pypi
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchmetrics              0.5.1                    pypi_0    pypi
[conda] torchvision               0.11.2               py37_cu102    pytorch

cc @ezyang @gchanan @zou3519 @bdhirsh @mruberry @peterbell10

**Comment by malfet** [comment]
I think numerical accuracy is fine (it becomes much lower if one to use `torch.cdouble` aka `torch.complex128` dtype)
What's problematic here is that spectral functions assume (but never check) that input tensor is contiguous in
https://github.com/pytorch/pytorch/blob/3ed27a96ed8031392b80bda6f80220c907b8336e/aten/src/ATen/native/mkl/SpectralOps.cpp#L421-L422

**Comment by peterbell10** [comment]
I can't reproduce your error numbers locally, only in colab. So, this is probably dependent on MKL version and possibly what CPU you're using. However, from colab runs I don't see any evidence of a bug. `x1.abs().sum()` is 35361 so the fractional error is only `1e-6` which is pretty close to the float epsilon. What probably happens is MKL is choosing a slightly different algorithm which results in the same answer with different rounding.

> These experiments show that the norm argument and dtype of the tensor impact the magnitude of the error. [...]
> It is a bit unclear why different normalization methods would impact the error, as the normalization constant should be the same regardless of if the tensor if permuted or not.

Norm is effectively multiplying `x1` and `x2` by the same constant, so I would expect the error to increases proportionally to the constant. To remove this effect you can simple divide by `x1.abs().sum()` and see that norm has no effect:
```
norm=forward, dtype=torch.complex64 error:  tensor(1.2532e-07)
norm=ortho, dtype=torch.complex64 error:  tensor(1.3712e-07)
norm=backward, dtype=torch.complex64 error:  tensor(1.3635e-07)
norm=forward, dtype=torch.complex128 error:  tensor(3.7589e-16, dtype=torch.float64)
norm=ortho, dtype=torch.complex128 error:  tensor(3.8273e-16, dtype=torch.float64)
norm=backward, dtype=torch.complex128 error:  tensor(3.8192e-16, dtype=torch.float64)
```

>I think numerical accuracy is fine (it becomes much lower if one to use torch.cdouble aka torch.complex128 dtype)
What's problematic here is that spectral functions assume (but never check) that input tensor is contiguous in

The strides are handled in `_plan_mkl_fft`, specifically here:
https://github.com/pytorch/pytorch/blob/3ed27a96ed8031392b80bda6f80220c907b8336e/aten/src/ATen/native/mkl/SpectralOps.cpp#L382-L391

**Comment by ngimel** [comment]
Removing hi priority per @peterbell10's comment, this is expected behavior.
